sootup.java.bytecode/src/main/java/sootup/java/bytecode/interceptors/typeresolving/TypePromotionVisitor.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ * #L%
+ */
+
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sootup.core.jimple.basic.Local;
+import sootup.core.jimple.basic.Value;
+import sootup.core.jimple.common.stmt.Stmt;
+import sootup.core.model.Body;
+import sootup.core.types.PrimitiveType;
+import sootup.core.types.Type;
+import sootup.java.bytecode.interceptors.typeresolving.types.AugIntegerTypes;
+
+public class TypePromotionVisitor extends TypeChecker {
+
+  private boolean failed = false;
+  private boolean typingChanged = true;
+
+  private static final Logger logger = LoggerFactory.getLogger(TypePromotionVisitor.class);
+
+  public TypePromotionVisitor(
+      @Nonnull Body.BodyBuilder builder,
+      @Nonnull AugEvalFunction evalFunction,
+      @Nonnull BytecodeHierarchy hierarchy) {
+    super(builder, evalFunction, hierarchy);
+  }
+
+  public Typing getPromotedTyping(Typing typing) {
+    setTyping(typing);
+    this.failed = false;
+    while (typingChanged && !failed) {
+      this.typingChanged = false;
+      for (Stmt stmt : builder.getStmts()) {
+        stmt.accept(this);
+      }
+    }
+    if (failed) {
+      return null;
+    }
+    return getTyping();
+  }
+
+  public static boolean isIntermediateType(Type type) {
+    return type.equals(AugIntegerTypes.getInteger1())
+        || type.equals(AugIntegerTypes.getInteger127())
+        || type.equals(AugIntegerTypes.getInteger32767());
+  }
+
+  public void visit(@Nonnull Value value, @Nonnull Type stdType, @Nonnull Stmt stmt) {
+    AugEvalFunction evalFunction = getFuntion();
+    BytecodeHierarchy hierarchy = getHierarchy();
+    Typing typing = getTyping();
+    Type evaType = evalFunction.evaluate(typing, value, stmt, graph);
+    if (evaType.equals(stdType)) {
+      return;
+    }
+    if (!hierarchy.isAncestor(stdType, evaType)) {
+      logger.error(
+          stdType
+              + " is not compatible with the value '"
+              + value
+              + "' in the statement: '"
+              + stmt
+              + "' !");
+      this.failed = true;
+    } else if (value instanceof Local && isIntermediateType(evaType)) {
+      Local local = (Local) value;
+      Type promotedType = promote(evaType, stdType);
+      if (promotedType != null && !promotedType.equals(evaType)) {
+        typing.set(local, promotedType);
+        this.typingChanged = true;
+      }
+    }
+  }
+
+  private Type promote(Type low, Type high) {
+    if (low instanceof AugIntegerTypes.Integer1Type) {
+      if (high instanceof PrimitiveType.IntType) {
+        return AugIntegerTypes.getInteger127();
+      } else if (high instanceof PrimitiveType.ShortType) {
+        return PrimitiveType.getByte();
+      } else if (high instanceof PrimitiveType.IntType) {
```

Comment:
yeah here we should take a look - there were a lot of highlights from IntelliJ

Reply:
fixed the duplicate type by looking at the promote method in old soot 😄 [here](https://github.com/soot-oss/soot/blob/d996672d8772df9654f23855ab21a10409e8c0ac/src/main/java/soot/jimple/toolkits/typing/fast/TypePromotionUseVisitor.java#L59)

branches were unreachable because instanceof is also true for superclasses.
Since Short is a Subclass of Integer it could not be reached.

I have replaced "x instanceof y" with "x.getClass() == y.class"